doc/decisions/base_name.md

```diff
@@ -0,0 +1,33 @@
+# Base Name
+
+## Problem
+
+Both applications and configuration file formats might have arbitrary strings which need to be
+encoded within a key name part.
+
+## Constraints
+
+- keySetBaseName (key, keyBaseName (key)) should be a NOP
+
+## Assumptions
+
+## Considered Alternatives
+
+- restrict what `keySetBaseName` can accept is against:
+  - easy to use of API: applications with some key names would suddenly fail
+  - hard to use it wrong API: storage plugins that do not pre-process key names properly would fail to build up a KeySet
+    which would be against our goal to support any configuration file format (e.g. TOML supports empty key part names).
+
+## Decision
+
+`keySetBaseName` must be able to encode any string as base name.
```

Comment:
> The whole decision becomes invalid, as soon as you add a keySetBaseNameFoo that does accept arbitrary strings and encodes them in such a way that keySetBaseName accepts them.

This solution would be less easy to use, and as such less preferable.

Reply:
> This solution would be less easy to use, and as such less preferable.

I don't agree, but okay.

Reply:
I now changed the argument in 5f524cf1a4c5da2d66e9e9eb7070976ddfa3339b to "seems to be a too big change in the storage plugins". (as you once said you can do the change in the core but not in the storage plugins.) Do we agree here?

Did you also think about providing a keyAddUnescapedName? This might be even more handy for property storage plugins like mini/simpleini: they could pre-process their key name to change the hierarchy separator to a NULL-byte and then call keyAddUnescapedName. Once we have something like this, `keySetBaseNameFoo` wouldn't be an option anymore, would it?

What I am mostly missing is some compelling use case where we really need some semantics in the key names. (Something which cannot be done in storage plugins nor with meta-data.) This would probably convince me to make the change.

Reply:
> I now changed the argument in [5f524cf](https://github.com/ElektraInitiative/libelektra/commit/5f524cf1a4c5da2d66e9e9eb7070976ddfa3339b) to "seems to be a too big change in the storage plugins". (as you once said you can do the change in the core but not in the storage plugins.) Do we agree here?

The specific proposal I made, was too much for me alone to implement it in all storage plugins. That was mainly, because it would take me a long time to familiarize myself with all the storage plugins. It is also not a good idea, if I implement all the changes. Take `toml` for example. I assume bauhaus93 will be its maintainer going forward. If I know implement some changes to make it work with a new `keySet*Name*` function, I will just make it harder for bauhaus93 to further maintain the plugin. Instead bauhaus93 should implement the changes.
Another reason, was if I am the only one working with the new API, we might miss some problem that someone else might spot. In general, as many people as possible should work with such a core API to ensure the API works as intended.

Apart from that, another proposal might be less work to implement. The whole thing also very much depends on the timeline. If there is no deadline, I don't mind changing all storage plugins. It might take me a year to do it and I would still prefer to have some help, but I would do it alone as well.

> Did you also think about providing a keyAddUnescapedName?

I didn't think about it no. But it also wouldn't solve anything, since it is essentially the same as calling `keyAddBaseName` repeatedly (just a little more efficient).

What I did think about however, was a function to construct `Key`s with an unescaped name. I didn't find a suitably usable API for it though. It also wouldn't have any benefit apart from performance, so I didn't invest too much time.

> What I am mostly missing is some compelling use case where we really need some semantics in the key names. (Something which cannot be done in storage plugins nor with meta-data.) This would probably convince me to make the change.

I don't think there is such a use-case. But I also don't think such a use-case is needed. It's a bit like saying:
Why use anything other then [Brainfuck](https://en.wikipedia.org/wiki/Brainfuck)? It's Turing-complete and unlike assembler it is not specific to a machine architecture, so it's enough to write all software.

Similarly, you could say: Why use Elektra, when you can write it yourself?

It's not about the use-case it is about usability and ease-of-use. To me at least, the current key names make things harder for everyone, just to make things easier for a handful of people.

Reply:
> I will just make it harder for bauhaus93 to further maintain the plugin. Instead bauhaus93 should implement the changes.

This sounds more like an excuse :wink:

We can discuss a shared implementation today.

> Another reason, was if I am the only one working with the new API, we might miss some problem that someone else might spot. In general, as many people as possible should work with such a core API to ensure the API works as intended.

I agree on that point. Without actually using the API it is not easy to see flaws.

>  I don't mind changing all storage plugins. It might take me a year to do it and I would still prefer to have some help, but I would do it alone as well.

With this attitude it will get done and you will get help.

> It's not about the use-case it is about usability and ease-of-use. To me at least, the current key names make things harder for everyone, just to make things easier for a handful of people.

If it makes something easier, I **very much** welcome the change. Somehow the point of **what** gets easier did not come across to me.